guava/src/com/google/common/graph/StandardMutableValueGraph.java

```diff
@@ -136,17 +137,23 @@ public boolean removeNode(N node) {
       }
     }
 
-    for (N successor : connections.successors()) {
+    // Since views are returned, we need to copy the successors that will be removed.
+    // Thus we avoid modifying the underlying view while iterating over it.
+    for (N successor : ImmutableSet.copyOf(connections.successors())) {
       // requireNonNull is safe because the node is a successor.
       requireNonNull(nodeConnections.getWithoutCaching(successor)).removePredecessor(node);
+      requireNonNull(connections.removeSuccessor(successor));
```

Comment:
It's been a while since I've looked at this code, but I'm a bit suspicious of the three different ways in which we're now accessing the connections of a node in this method:

* direct reference (or alias) for the internal data structure: `nodeConnections.get(node)` (via `connections`, line 127; `nodeConnections` is declared in the superclass as `MapIteratorCache<N, GraphConnections<N, V>> nodeConnections`)
* direct reference without caching: `nodeConnections.getWithoutCaching()` (line 144)
* (immutable) copy of the output of that structure (line 142)

We should (a) validate that we want those three different mechanisms, or use only the ones we actually want, and (b) document which one(s) we do want, and why. 

There's a lot of clever stuff going on in this code, which makes it easy to make mistakes in updating it.

(To be clear, I think you've explained why you want the copy of the set in the places that you've made it, but the reason why we're using each of the above two mechanisms in different places is less clear.)